File: 4_Sorting_Algorithms/Quick_Sort/C++/quick_sort.cpp
#include "quick_sort.hpp"

/**
 * quick_sort - Sorts an array using the quick sort algorithm.
 *
 * @arr: Reference to the array to be sorted.
 */
std::vector<int> quick_sort(std::vector<int> arr) {
    if (arr.size() < 2){
        return arr;
    }

    int pivot {arr.back()};
    std::vector<int> left {};
    std::vector<int> right {};

    for (size_t i = 0; i < arr.size() - 1; ++i) {
        if (arr[i] < pivot) {
            left.push_back(arr[i]);
        } else {
            right.push_back(arr[i]);
        }
    }

    std::vector<int> solution = quick_sort(left);
    solution.push_back(pivot);
    std::vector<int> right_sorted = quick_sort(right);
    solution.insert(solution.end(), right_sorted.begin(), right_sorted.end());

    return solution;
}

File: 4_Sorting_Algorithms/Quick_Sort/Python/quick_sort.py
"""
This is the merge sort library
"""
from typing import List


def quick_sort(arr: List[int]) -> List[int]:
    if len(arr) < 2:
        return arr

    pivot = arr[-1]
    left = []
    right = []

    for i in range(len(arr) - 1):
        if arr[i] < pivot:
            left.append(arr[i])
        else:
            right.append(arr[i])

    return quick_sort(left) + [pivot] + quick_sort(right)
